File: integrations/strands/tests/test_telemetry.py
"""
Tests for telemetry module — event mapping and JSONL emission shape.

Uses stdlib unittest only; no strands-agents required.
"""

import json
import os
import tempfile
import unittest
from pathlib import Path


class TestStrandsToCanonicalMapping(unittest.TestCase):
    """Verify the Strands → canonical event-name mapping table."""

    def setUp(self):
        from flow_agents_strands.telemetry import STRANDS_TO_CANONICAL
        self.mapping = STRANDS_TO_CANONICAL

    def test_all_expected_keys_present(self):
        expected = {
            "AgentInitializedEvent",
            "BeforeInvocationEvent",
            "AfterInvocationEvent",
            "BeforeToolCallEvent",
            "AfterToolCallEvent",
            "AfterModelCallEvent",
            "MessageAddedEvent",
        }
        self.assertEqual(expected, set(self.mapping.keys()))

    def test_before_invocation_maps_to_user_prompt_submit(self):
        self.assertEqual("userPromptSubmit", self.mapping["BeforeInvocationEvent"])

    def test_after_invocation_maps_to_stop(self):
        self.assertEqual("stop", self.mapping["AfterInvocationEvent"])

    def test_before_tool_call_maps_to_pre_tool_use(self):
        self.assertEqual("preToolUse", self.mapping["BeforeToolCallEvent"])

    def test_after_tool_call_maps_to_post_tool_use(self):
        self.assertEqual("postToolUse", self.mapping["AfterToolCallEvent"])

    def test_agent_initialized_maps_to_agent_spawn(self):
        self.assertEqual("agentSpawn", self.mapping["AgentInitializedEvent"])

    def test_all_values_are_strings(self):
        for k, v in self.mapping.items():
            with self.subTest(key=k):
                self.assertIsInstance(v, str)


class TestTelemetrySinkEmission(unittest.TestCase):
    """Verify JSONL emission shape matches the canonical Flow Agents schema."""

    def setUp(self):
        from flow_agents_strands.telemetry import TelemetrySink
        self._tmp = tempfile.TemporaryDirectory()
        self._sink_dir = Path(self._tmp.name)
        self._sink = TelemetrySink(
            sink_path=str(self._sink_dir),
            agent_name="test-agent",
            runtime="strands-test",
        )

    def tearDown(self):
        self._tmp.cleanup()

    def _read_events(self):
        log_file = self._sink_dir / "full.jsonl"
        if not log_file.exists():
            return []
        lines = log_file.read_text(encoding="utf-8").strip().splitlines()
        return [json.loads(line) for line in lines if line.strip()]

    def test_session_start_event_shape(self):
        evt = self._sink.emit_session_start()

        # Top-level required fields (mirrors build_base_event in telemetry.sh)
        self.assertEqual("0.3.0", evt["schema_version"])
        self.assertIn("timestamp", evt)
        self.assertIn("session_id", evt)
        self.assertIn("event_id", evt)
        self.assertEqual("session.start", evt["event_type"])

        # Agent sub-object
        agent = evt["agent"]
        self.assertEqual("test-agent", agent["name"])
        self.assertEqual("strands-test", agent["runtime"])

    def test_session_start_written_to_jsonl(self):
        self._sink.emit_session_start()
        events = self._read_events()
        self.assertEqual(1, len(events))
        self.assertEqual("session.start", events[0]["event_type"])

    def test_tool_invoke_event_shape(self):
        evt = self._sink.emit_tool_invoke("edit", {"path": "foo.py"})
        self.assertEqual("tool.invoke", evt["event_type"])
        self.assertEqual("edit", evt["tool"]["name"])
        self.assertEqual("fs_write", evt["tool"]["normalized_name"])
        self.assertEqual({"path": "foo.py"}, evt["tool"]["input"])

    def test_tool_result_event_shape(self):
        evt = self._sink.emit_tool_result("read", "file contents")
        self.assertEqual("tool.result", evt["event_type"])
        self.assertEqual("read", evt["tool"]["name"])
        self.assertEqual("fs_read", evt["tool"]["normalized_name"])
        self.assertEqual("file contents", evt["tool"]["output"])

    def test_session_end_event_shape(self):
        evt = self._sink.emit_session_end(duration_s=42.5)
        self.assertEqual("session.end", evt["event_type"])
        self.assertAlmostEqual(42.5, evt["session"]["duration_s"])

    def test_user_prompt_submit_event_shape(self):
        evt = self._sink.emit("userPromptSubmit")
        self.assertEqual("turn.user", evt["event_type"])

    def test_hook_context_present(self):
        """Every event must include a hook sub-object (mirrors add_hook_context)."""
        evt = self._sink.emit_session_start()
        self.assertIn("hook", evt)
        hook = evt["hook"]
        self.assertIn("event_name", hook)
        self.assertIn("source", hook)
        self.assertEqual("strands", hook["source"])

    def test_multiple_events_same_session_id(self):
        self._sink.emit_session_start()
        self._sink.emit_tool_invoke("read", {})
        self._sink.emit_session_end()
        events = self._read_events()
        session_ids = {e["session_id"] for e in events}
        self.assertEqual(1, len(session_ids), "All events must share one session_id")

    def test_jsonl_each_line_valid_json(self):
        self._sink.emit_session_start()
        self._sink.emit_tool_invoke("bash", {"command": "ls"})
        self._sink.emit_session_end(duration_s=1.0)
        log_file = self._sink_dir / "full.jsonl"
        for line in log_file.read_text(encoding="utf-8").splitlines():
            if line.strip():
                parsed = json.loads(line)   # will raise on invalid JSON
                self.assertIsInstance(parsed, dict)

    def test_sink_path_directory_creates_full_jsonl(self):
        """When sink_path is a directory, file is named full.jsonl."""
        from flow_agents_strands.telemetry import TelemetrySink
        with tempfile.TemporaryDirectory() as d:
            sink = TelemetrySink(sink_path=d)
            sink.emit_session_start()
            log_file = Path(d) / "full.jsonl"
            self.assertTrue(log_file.exists())

    def test_emit_steering_event_type(self):
        evt = self._sink.emit_steering("STATE: task is status:in_progress")
        self.assertEqual("turn.user", evt["event_type"])
        self.assertIn("steering_context", evt["turn"])


class TestNormalizeToolName(unittest.TestCase):
    """Spot-check normalize_tool_name mirrors telemetry.sh."""

    def setUp(self):
        from flow_agents_strands.telemetry import _normalize_tool_name
        self._fn = _normalize_tool_name

    def test_bash(self):
        self.assertEqual("execute_bash", self._fn("bash"))

    def test_edit_is_fs_write(self):
        self.assertEqual("fs_write", self._fn("edit"))

    def test_read_is_fs_read(self):
        self.assertEqual("fs_read", self._fn("read"))

    def test_unknown_passthrough(self):
        self.assertEqual("my_custom_tool", self._fn("my_custom_tool"))


if __name__ == "__main__":
    unittest.main()

File: integrations/strands-ts/test/test-telemetry.ts
/**
 * test-telemetry.ts — Tests for telemetry module.
 *
 * Covers: event mapping, JSONL emission shape, normalizeToolName.
 * Uses node:test only — no additional dependencies.
 */

import { test, describe } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { TelemetrySink, STRANDS_TO_CANONICAL, normalizeToolName, SCHEMA_VERSION } from "../src/telemetry.js";

// ---------------------------------------------------------------------------
// STRANDS_TO_CANONICAL mapping table
// ---------------------------------------------------------------------------

describe("STRANDS_TO_CANONICAL mapping", () => {
  test("contains all expected Strands TS event class names", () => {
    const expected = new Set([
      "BeforeInvocationEvent",
      "AfterInvocationEvent",
      "BeforeToolCallEvent",
      "AfterToolCallEvent",
      "AgentInitializedEvent",
      "AfterModelCallEvent",
      "MessageAddedEvent",
    ]);
    assert.deepStrictEqual(new Set(Object.keys(STRANDS_TO_CANONICAL)), expected);
  });

  test("BeforeInvocationEvent → userPromptSubmit", () => {
    assert.strictEqual(STRANDS_TO_CANONICAL.BeforeInvocationEvent, "userPromptSubmit");
  });

  test("AfterInvocationEvent → stop", () => {
    assert.strictEqual(STRANDS_TO_CANONICAL.AfterInvocationEvent, "stop");
  });

  test("BeforeToolCallEvent → preToolUse", () => {
    assert.strictEqual(STRANDS_TO_CANONICAL.BeforeToolCallEvent, "preToolUse");
  });

  test("AfterToolCallEvent → postToolUse", () => {
    assert.strictEqual(STRANDS_TO_CANONICAL.AfterToolCallEvent, "postToolUse");
  });

  test("AgentInitializedEvent → agentSpawn", () => {
    assert.strictEqual(STRANDS_TO_CANONICAL.AgentInitializedEvent, "agentSpawn");
  });

  test("all values are non-empty strings", () => {
    for (const [key, value] of Object.entries(STRANDS_TO_CANONICAL)) {
      assert.strictEqual(typeof value, "string", `${key} value must be a string`);
      assert.ok(value.length > 0, `${key} value must be non-empty`);
    }
  });
});

// ---------------------------------------------------------------------------
// normalizeToolName
// ---------------------------------------------------------------------------

describe("normalizeToolName", () => {
  test("bash → execute_bash", () => {
    assert.strictEqual(normalizeToolName("bash"), "execute_bash");
  });

  test("edit → fs_write", () => {
    assert.strictEqual(normalizeToolName("edit"), "fs_write");
  });

  test("write → fs_write", () => {
    assert.strictEqual(normalizeToolName("write"), "fs_write");
  });

  test("read → fs_read", () => {
    assert.strictEqual(normalizeToolName("read"), "fs_read");
  });

  test("task → use_subagent", () => {
    assert.strictEqual(normalizeToolName("task"), "use_subagent");
  });

  test("unknown passthrough", () => {
    assert.strictEqual(normalizeToolName("my_custom_tool"), "my_custom_tool");
  });
});

// ---------------------------------------------------------------------------
// TelemetrySink emission shape
// ---------------------------------------------------------------------------

function makeTempSink(agentName = "test-agent", runtime = "strands-test"): { sink: TelemetrySink; dir: string } {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "fa-ts-telemetry-"));
  const sink = new TelemetrySink({ sinkPath: dir, agentName, runtime });
  return { sink, dir };
}

function readEvents(dir: string): Record<string, unknown>[] {
  const logFile = path.join(dir, "full.jsonl");
  if (!fs.existsSync(logFile)) return [];
  const lines = fs.readFileSync(logFile, "utf8").trim().split("\n");
  return lines.filter((l) => l.trim()).map((l) => JSON.parse(l) as Record<string, unknown>);
}

describe("TelemetrySink emission shape", () => {
  test("session start event has correct schema_version and event_type", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitSessionStart();
    assert.strictEqual(evt.schema_version, SCHEMA_VERSION);
    assert.strictEqual(evt.event_type, "session.start");
    // cleanup
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("session start written to JSONL", () => {
    const { sink, dir } = makeTempSink();
    sink.emitSessionStart();
    const events = readEvents(dir);
    assert.strictEqual(events.length, 1);
    assert.strictEqual((events[0] as Record<string, unknown>).event_type, "session.start");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("event has required top-level fields (timestamp, session_id, event_id)", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitSessionStart();
    assert.ok("timestamp" in evt, "missing timestamp");
    assert.ok("session_id" in evt, "missing session_id");
    assert.ok("event_id" in evt, "missing event_id");
    assert.ok("agent" in evt, "missing agent");
    assert.ok("hook" in evt, "missing hook");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("agent sub-object has name and runtime", () => {
    const { sink, dir } = makeTempSink("my-agent", "strands-ts");
    const evt = sink.emitSessionStart();
    const agent = evt.agent as Record<string, unknown>;
    assert.strictEqual(agent.name, "my-agent");
    assert.strictEqual(agent.runtime, "strands-ts");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("hook sub-object has event_name and source fields", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitSessionStart();
    const hook = evt.hook as Record<string, unknown>;
    assert.ok("event_name" in hook, "hook.event_name missing");
    assert.ok("source" in hook, "hook.source missing");
    assert.strictEqual(hook.source, "strands-ts");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("tool invoke event has correct event_type and tool fields", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitToolInvoke("edit", { path: "foo.py" });
    assert.strictEqual(evt.event_type, "tool.invoke");
    const tool = evt.tool as Record<string, unknown>;
    assert.strictEqual(tool.name, "edit");
    assert.strictEqual(tool.normalized_name, "fs_write");
    assert.deepStrictEqual(tool.input, { path: "foo.py" });
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("tool result event has correct event_type", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitToolResult("read", "file contents");
    assert.strictEqual(evt.event_type, "tool.result");
    const tool = evt.tool as Record<string, unknown>;
    assert.strictEqual(tool.name, "read");
    assert.strictEqual(tool.normalized_name, "fs_read");
    assert.strictEqual(tool.output, "file contents");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("session end has event_type session.end", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitSessionEnd(1000);
    assert.strictEqual(evt.event_type, "session.end");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("user prompt submit has event_type turn.user", () => {
    const { sink, dir } = makeTempSink();
    const evt = sink.emitUserPromptSubmit();
    assert.strictEqual(evt.event_type, "turn.user");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("multiple events share the same session_id", () => {
    const { sink, dir } = makeTempSink();
    sink.emitSessionStart();
    sink.emitToolInvoke("read", {});
    sink.emitSessionEnd();
    const events = readEvents(dir);
    const sessionIds = new Set(events.map((e) => (e as Record<string, unknown>).session_id));
    assert.strictEqual(sessionIds.size, 1, "All events must share one session_id");
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("each JSONL line is valid JSON", () => {
    const { sink, dir } = makeTempSink();
    sink.emitSessionStart();
    sink.emitToolInvoke("bash", { command: "ls" });
    sink.emitSessionEnd(500);
    const logFile = path.join(dir, "full.jsonl");
    const lines = fs.readFileSync(logFile, "utf8").split("\n");
    for (const line of lines) {
      if (!line.trim()) continue;
      const parsed = JSON.parse(line); // throws on invalid JSON
      assert.strictEqual(typeof parsed, "object");
    }
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test("sinkPath directory creates full.jsonl", () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), "fa-ts-sink-dir-"));
    const sink = new TelemetrySink({ sinkPath: dir });
    sink.emitSessionStart();
    assert.ok(fs.existsSync(path.join(dir, "full.jsonl")));
    fs.rmSync(dir, { recursive: true, force: true });
  });
});
